scripts/build/termux_step_massage.sh

```diff
@@ -205,7 +206,8 @@ termux_step_massage() {
 		local nproc=$(nproc)
 		echo "INFO: Identifying files with nproc=${nproc}"
 		local t0=$(get_epoch)
-		local files=$(find . -type f)
+		# Match all files (including binaries) containing a ELF magic at the start of any line.
+		local files="$(grep --recursive --binary-files=text --files-with-matches --max-count=1 $'^\x7fELF' .)"
```

Comment:
I know we wanna limit the refactoring in this PR, but I think renaming this variable to `elf_files` helps get its purpose across.

```suggestion
		local elf_files="$(grep --recursive --binary-files=text --files-with-matches --max-count=1 $'^\x7fELF' .)"
```

The comment is fine, but shouldn't we be checking *specifically* the start of the file?
Since `\x7fELF` needs to be the first 4 bytes of the file.
https://en.wikipedia.org/wiki/Executable_and_Linkable_Format#ELF_header

Reply:
I'd like to not touch other code in this script, limiting it to two very specific changes if you do not mind.

I did not find any way to make grep limit search only to the first line without additional overhead. Since we already optimized file searching (under 0.1% of original time, making `TERMUX_PKG_NO_SHEBANG_FIX` variable existance kinda pointless since it does not cause original overhead) and the delay caused by the code is so minimal you can hardly notice it, I'd like not not dive deeper into this rabbit hole.

Reply:
I plan to come back to refactoring later since I already see a lot of other places we can improve too.

Reply:
Sounds good to me.

Reply:
The `$'^\x7fELF'` regex does not seem to work properly, it matches some static libraries on my system too.

Better to loop on `find -type f .` and manually create a file list by checking if file matches `hexdump -n 4 -e '4/1 "%2x" " "' -- "$file_path" | grep -q "^7f454c46"`.

```shell
file_path=/lib/x86_64-linux-gnu/libselinux.a
$ hexdump -n 4 -e '4/1 "%2x" " "' -- "$file_path"
213c6172
$ grep --recursive --binary-files=text --files-with-matches --max-count=1 $'^\x7fELF' $file_path
/lib/x86_64-linux-gnu/libselinux.a
```

Reply:
> it matches some static libraries on my system too that do not start with ELF header.

That is a good point, and actually, I noticed that too, BUT I thought it was considered expected behavior, for only the reason that, if the regex there were to completely exclude _all_ `.a` files, then this line would become unreachable code, right?

https://github.com/termux/termux-packages/blob/2a8ed11de50a9b3a4e8392efc828043c220edac6/scripts/build/termux_step_massage.sh#L246

If all files that are not true ELFs will be excluded early, then some other parts of the function seem to become unreachable code.

Reply:
> The `$'^\x7fELF'` regex does not seem to work properly, it matches some static libraries on my system too that do not start with ELF header.

As stated in code it does not filter files STARTING with elf magic. It finds files files containing elf magic at the start of any line. It is a basic filtration to slightly (actually almost completely) reduce amount of `fork`+`execve`+linking+executable (`readelf`) initializing+ etc... invocations overhead.
Final filtering is happening in `readelf` as it happened later.

Reply:
@agnostic-apollo I started doubting about this solution after your words and checked it.
```
$ time first="$(grep --recursive --binary-files=text --files-with-matches --max-count=1 $'^\x7fELF' /data | xargs -P"$(nproc)" -i bash -c 'if $(command -v readelf) -h "{}" &>/dev/null; then echo "{}"; fi' | sort)"
 
real    0m0.288s
user    0m0.719s
sys     0m1.002s
$ time second="$(find -L /data -type f | xargs -P"$(nproc)" -i bash -c 'if $(command -v readelf) -h "{}" &>/dev/null; then echo "{}"; fi' | sort)"

real    0m27.440s
user    1m31.870s
sys     2m6.035s
```
`$first` and `$second` are different, but that happens because the `grep` solution automatically strips all symlink and that reduces amount of files to be checked even more.

I checked it with 
```bash
$ time first="$(grep --recursive --binary-files=text --files-with-matches --max-count=1 $'^\x7fELF' /data | xargs -P"$(nproc)" -i bash -c 'if $(command -v readelf) -h "{}" &>/dev/null; then echo "{}"; fi' | sort)" 
real    0m0.286s
user    0m0.736s
sys     0m0.998s

# As you can see there is no `-L` argument in `find` so we ignore symlinks.
$ time second="$(find /data -type f | xargs -P"$(nproc)" -i bash -c 'if $(command -v readelf) -h "{}" &>/dev/null; then echo "{}"; fi' | sort)"
real    0m10.638s
user    0m35.362s
sys     0m48.428s

$ [[ "${first}" == "${second}" ]] && echo "identical" || echo "different"
identical
```

Results are identical but the whole process takes less time, so prefiltering works fine.

Reply:
@agnostic-apollo Just to make sure my code does not have any differences with old behaviour I installed all possible packages to termux-docker container with 
```
apt update
for pkg in $(apt-cache pkgnames); do
  sudo DEBIAN_FRONTEND=noninteractive apt install -y $pkg || continue
done
```
and tried to check perfomance boost and if there are any differences in output.
Unfortunately it had some differences.
```
$ time first="$(grep --recursive --binary-files=text --files-with-matches --max-count=1 $'^\x7fELF' /data | xargs -P"$(nproc)" -i bash -c 'if $(command -v readelf) -h "{}" &>/dev/null; then echo "{}"; fi' | sort)"
real    0m5.761s
user    0m31.031s
sys     0m41.437s

$ time second="$(find /data -type f | xargs -P"$(nproc)" -i bash -c 'if $(command -v readelf) -h "{}" &>/dev/null; then echo "{}"; fi' | sort)"
real    2m29.109s
user    17m46.157s
sys     25m30.750s

~ $ diff -Naur <(echo "${first}") <(echo "${second}")
--- /proc/self/fd/63    2025-04-28 06:50:00.388488728 +0000
+++ /proc/self/fd/62    2025-04-28 06:50:00.388488728 +0000
@@ -2848,6 +2848,7 @@
 /data/data/com.termux/files/usr/glibc/lib/libXtst.so.6.1.0
 /data/data/com.termux/files/usr/glibc/lib/libXxf86vm.so.1.0.0
 /data/data/com.termux/files/usr/glibc/lib/libacl.so.1.1.2301
+/data/data/com.termux/files/usr/glibc/lib/libanl.a
 /data/data/com.termux/files/usr/glibc/lib/libanl.so.1
 /data/data/com.termux/files/usr/glibc/lib/libaom.so.3.12.0
 /data/data/com.termux/files/usr/glibc/lib/libapr-1.so.0.7.5
@@ -2893,6 +2894,7 @@
 /data/data/com.termux/files/usr/glibc/lib/libdbus-1.so.3.37.0
 /data/data/com.termux/files/usr/glibc/lib/libdconf.so.1.0.0
 /data/data/com.termux/files/usr/glibc/lib/libdebuginfod-0.192.so
+/data/data/com.termux/files/usr/glibc/lib/libdl.a
 /data/data/com.termux/files/usr/glibc/lib/libdl.so.2
 /data/data/com.termux/files/usr/glibc/lib/libdrm.so.2.124.0
 /data/data/com.termux/files/usr/glibc/lib/libdrm_amdgpu.so.1.124.0
@@ -3034,6 +3036,7 @@
 /data/data/com.termux/files/usr/glibc/lib/libpq.so.5.17
 /data/data/com.termux/files/usr/glibc/lib/libpsl.so.5.3.5
 /data/data/com.termux/files/usr/glibc/lib/libpsx.so.2.69
+/data/data/com.termux/files/usr/glibc/lib/libpthread.a
 /data/data/com.termux/files/usr/glibc/lib/libpthread.so.0
 /data/data/com.termux/files/usr/glibc/lib/libquadmath.a
 /data/data/com.termux/files/usr/glibc/lib/libquadmath.so.0.0.0
@@ -3041,6 +3044,7 @@
 /data/data/com.termux/files/usr/glibc/lib/libresolv.a
 /data/data/com.termux/files/usr/glibc/lib/libresolv.so.2
 /data/data/com.termux/files/usr/glibc/lib/librhash.so.1.4.5
+/data/data/com.termux/files/usr/glibc/lib/librt.a
 /data/data/com.termux/files/usr/glibc/lib/librt.so.1
 /data/data/com.termux/files/usr/glibc/lib/libserf-1.so.1.3.0
 /data/data/com.termux/files/usr/glibc/lib/libsframe.a
@@ -3069,6 +3073,7 @@
 /data/data/com.termux/files/usr/glibc/lib/libunwind-setjmp.so.0.0.0
 /data/data/com.termux/files/usr/glibc/lib/libunwind-x86.so.8.1.0
 /data/data/com.termux/files/usr/glibc/lib/libunwind.so.8.1.0
+/data/data/com.termux/files/usr/glibc/lib/libutil.a
 /data/data/com.termux/files/usr/glibc/lib/libutil.so.1
 /data/data/com.termux/files/usr/glibc/lib/libuuid.so.1.3.0
 /data/data/com.termux/files/usr/glibc/lib/libva-drm.so.2.2200.0
@@ -4309,6 +4314,7 @@
 /data/data/com.termux/files/usr/lib/libQt5FbSupport.a
 /data/data/com.termux/files/usr/lib/libQt5FontDatabaseSupport.a
 /data/data/com.termux/files/usr/lib/libQt5Gui.so.5.15.16
+/data/data/com.termux/files/usr/lib/libQt5InputSupport.a
 /data/data/com.termux/files/usr/lib/libQt5Network.so.5.15.16
 /data/data/com.termux/files/usr/lib/libQt5PrintSupport.so.5.15.16
 /data/data/com.termux/files/usr/lib/libQt5SerialPort.so.5.15.16
@@ -6322,6 +6328,14 @@
 /data/data/com.termux/files/usr/lib/thunarx-3/thunar-sbr.so
 /data/data/com.termux/files/usr/lib/thunarx-3/thunar-uca.so
 /data/data/com.termux/files/usr/lib/thunarx-3/thunar-wallpaper-plugin.so
+/data/data/com.termux/files/usr/lib/tinygo/lib/wasi-libc/sysroot/lib/wasm32-wasi/libcrypt.a
+/data/data/com.termux/files/usr/lib/tinygo/lib/wasi-libc/sysroot/lib/wasm32-wasi/libdl.a
+/data/data/com.termux/files/usr/lib/tinygo/lib/wasi-libc/sysroot/lib/wasm32-wasi/libm.a
+/data/data/com.termux/files/usr/lib/tinygo/lib/wasi-libc/sysroot/lib/wasm32-wasi/libpthread.a
+/data/data/com.termux/files/usr/lib/tinygo/lib/wasi-libc/sysroot/lib/wasm32-wasi/libresolv.a
+/data/data/com.termux/files/usr/lib/tinygo/lib/wasi-libc/sysroot/lib/wasm32-wasi/librt.a
+/data/data/com.termux/files/usr/lib/tinygo/lib/wasi-libc/sysroot/lib/wasm32-wasi/libutil.a
+/data/data/com.termux/files/usr/lib/tinygo/lib/wasi-libc/sysroot/lib/wasm32-wasi/libxnet.a
 /data/data/com.termux/files/usr/lib/tumbler-1/plugins/cache/tumbler-xdg-cache.so
 /data/data/com.termux/files/usr/lib/tumbler-1/plugins/tumbler-cover-thumbnailer.so
 /data/data/com.termux/files/usr/lib/tumbler-1/plugins/tumbler-desktop-thumbnailer.so
```

After some investigation I found out that:
1. These static libraries did not contain ELF magic anywhere in the code, also `readelf -h` returns 0 with empty output.
2. Static libraries have different magic (`^!<arch>`).

Adding one more magic to the grep command made the search ~50% slower, but now behaviour completely matches master branch behaviour.
```
$ time first="$(grep --recursive --binary-files=text --files-with-matches --max-count=1 $'^\x7fELF' /data | xargs -P"$(nproc)" -i bash -c 'if $(command -v readelf) -h "{}" &>/dev/null; then echo "{}"; fi' | sort)"
real    0m5.778s
user    0m31.056s
sys     0m42.625s

$ time first="$(grep --recursive --binary-files=text --files-with-matches --max-count=1 -e $'^\x7fELF' -e '^!<arch>' /data | xargs -P"$(nproc)" -i bash -c 'if $(command -v readelf) -h "{}" &>/dev/null; then echo "{}"; fi' | sort)"
real    0m7.548s
user    0m35.310s
sys     0m44.582s

$ [[ "${first}" == "${second}" ]] && echo "identical" || echo "different"
identical
```

Thank you!

Reply:
Okay good. I don't really have the time to understand the complex symbol logic. Not sure if some other file have some different header and will get excluded, but I guess for now, can go ahead if others are fine with it who may no more.

Use `-e $'^((\x7fELF)|(!<arch>))'` for faster speed, adding another `-e` would spawn another regex engine.